docs/source/cuda/cuda_array_interface.rst

```diff
@@ -76,24 +76,376 @@ The following are optional entries:
+       start = cuda.grid(1)
+       stride = cuda.gridsize(1)
+       for i in range(start, x.shape[0], stride):
+           out[i] = x[i] + y[i]
+
+   a = cupy.arange(10)
+   b = a * 2
+   out = cupy.zeros_like(a)
+
+   add[1, 32](a, b, out)
+
+When the ``add`` kernel is launched:
+
+- ``a``, ``b``, ``out`` are Producers.
+- The ``add`` kernel is the Consumer.
+- The User Code is specifically ``add[1, 32](a, b, out)``.
+- The author of the code is the User.
+
+
+Design Motivations
+~~~~~~~~~~~~~~~~~~
+
+Elements of the CAI design related to synchronization seek to fulfill these
+requirements:
+
+1. Producers and Consumers that exchange data through the CAI must be able to do
+   so without data races.
+2. Requirement 1 should be met without requiring the user to be
+   aware of any particulars of the CAI - in other words, exchanging data between
+   Producers and Consumers that operate on data asynchronously should be correct
+   by default.
+
+   - An exception to this requirement is made for Producers and Consumers that
+     explicitly document that the User is required to take additional steps to
+     ensure correctness with respect to synchronization. In this case, Users
+     are required to understand the details of the CUDA Array Interface, and
+     the Producer/Consumer library documentation must specify the steps that
+     Users are required to take.
+
+     Use of this exception should be avoided where possible, as it is provided
+     for libraries that cannot implement the synchronization semantics without
+     the involvement of the User - for example, those interfacing with
+     third-party libraries oblivious to the CUDA Array Interface.
+
+3. Where the User is aware of the particulars of the CAI and implementation
+   details of the Producer and Consumer, they should be able to, at their
+   discretion, override some of the synchronization semantics of the interface
+   to reduce the synchronization overhead. Overriding synchronization semantics
+   implies that:
+
+   - The CAI design, and the design and implementation of the Producer and
+     Consumer do not specify or guarantee correctness with respect to data
+     races.
+   - Instead, the User is responsible for ensuring correctness with respect to
+     data races.
+
+
+Interface Requirements
+~~~~~~~~~~~~~~~~~~~~~~
+
+The ``stream`` entry enables Producers and Consumers to avoid hazards when
+exchanging data. Expected behaviour of the Consumer is as follows:
+
+* When ``stream`` is not present or is ``None``:
+
+  - No synchronization is required on the part of the Consumer.
+  - The Consumer may enqueue operations on the underlying data immediately on
+    any stream.
+
+* When ``stream`` is an integer, its value indicates the stream on which the
+  Producer may have in-progress operations on the data, and which the Consumer
+  is expected to either:
+
+  - Synchronize on before accessing the data, or
+  - Enqueue operations in when accessing the data.
```

Comment:
The CUDA documentation seems to talk about enqueuing operations in streams rather than on them (e.g. [`cudaStreamAttachMemAsync` description](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1gc3bb7ccb325219073183a629d7c2756a)) so "in" is consistent with that usage.

Reply:
ok, will leave as is, thanks for checking.